zipfs/file.go

```diff
@@ -0,0 +1,164 @@
+package zipfs
+
+import (
+	"archive/zip"
+	"io"
+	"os"
+	"path/filepath"
+	"syscall"
+
+	"github.com/spf13/afero"
+)
+
+type File struct {
+	fs            *Fs
+	zipfile       *zip.File
+	reader        io.ReadCloser
+	offset        int64
+	isdir, closed bool
+	buf           []byte
+}
+
+func (f *File) fillBuffer(offset int64) (err error) {
+	if f.reader == nil {
+		if f.reader, err = f.zipfile.Open(); err != nil {
+			return
+		}
+	}
+	if offset > int64(f.zipfile.UncompressedSize64) {
+		offset = int64(f.zipfile.UncompressedSize64)
+		err = io.EOF
+	}
+	if len(f.buf) >= int(offset) {
+		return
+	}
+	buf := make([]byte, int(offset)-len(f.buf))
+	n, _ := io.ReadFull(f.reader, buf)
+	if n > 0 {
+		f.buf = append(f.buf, buf[:n]...)
+	}
+	return
+}
+
+func (f *File) Close() (err error) {
+	f.zipfile = nil
+	f.closed = true
+	f.buf = nil
+	if f.reader != nil {
+		err = f.reader.Close()
+		f.reader = nil
+	}
+	return
+}
+
+func (f *File) Read(p []byte) (n int, err error) {
+	if f.isdir {
+		return 0, syscall.EISDIR
+	}
+	if f.closed {
+		return 0, afero.ErrFileClosed
+	}
+	err = f.fillBuffer(f.offset + int64(len(p)))
+	n = copy(p, f.buf[f.offset:])
+	f.offset += int64(len(p))
+	return
+}
+
+func (f *File) ReadAt(p []byte, off int64) (n int, err error) {
+	if f.isdir {
+		return 0, syscall.EISDIR
+	}
+	if f.closed {
+		return 0, afero.ErrFileClosed
+	}
+	err = f.fillBuffer(off + int64(len(p)))
+	n = copy(p, f.buf[int(off):])
+	return
+}
+
+func (f *File) Seek(offset int64, whence int) (int64, error) {
+	if f.isdir {
+		return 0, syscall.EISDIR
+	}
+	if f.closed {
+		return 0, afero.ErrFileClosed
+	}
+	switch whence {
+	case os.SEEK_SET:
+	case os.SEEK_CUR:
+		offset += f.offset
+	case os.SEEK_END:
+		offset += int64(f.zipfile.UncompressedSize64)
```

Comment:
Wont this cause an out of range error below, eg. if I want to move the offset 2 bytes before the end? Shouldn't this be:
```go
offset = int64(f.zipfile.UncompressedSize64) - offset
```
Or is offset meant to be provided as a negative number?

Reply:
> Or is offset meant to be provided as a negative number?
Exactly. See the _lseek(2)_ manpage on any Unix system.